lexicon/providers/dinahosting.py

```diff
@@ -0,0 +1,258 @@
+        data = {
+            'domain': self.domain,
+            'hostname': rel_name
+        }
+
+        if rtype == 'A':
+            data['command'] = "Domain_Zone_AddTypeA"
+            data['ip'] = content
+        elif rtype == 'AAAA':
+            data['command'] = "Domain_Zone_AddTypeAAAA"
+            data['ip'] = content
+        elif rtype == 'CNAME':
+            data['command'] = "Domain_Zone_AddTypeCname"
+            data['destinationHostname'] = content
+        elif rtype == 'TXT':
+            data['command'] = "Domain_Zone_AddTypeTXT"
+            data['text'] = content
+
+        try:
+            self._post('', data)
+        except requests.exceptions.HTTPError as err:
+            already_exists = ((err.response.json()['responseCode'] == RC_OBJECT_EXISTS) or
+                              (err.response.json()['responseCode'] == RC_COMMAND_FAILED))
+            if not already_exists:
+                raise
+
+        LOGGER.debug('create_record: %s', True)
+        return True
+
+    def _list_records(self, rtype=None, name=None, content=None):
+        if rtype:
+            formatted_rtype = "Cname" if rtype == 'CNAME' else rtype
+            data = {
+                'command': 'Domain_Zone_GetType{0}'.format(formatted_rtype),
+                'domain': self.domain
+            }
+        else:
+            data = {
+                'command': 'Domain_Zone_GetAll',
+                'domain': self.domain
+            }
+
+        payload = self._post('', data)
+
+        #content_obj = self._parse_record_content(rtype, content)
+        #content = content_obj['data']
+
+        records = []
+        for record in payload['data']:
+            raw_type = self._parse_record_type(record)
+            raw_content = self._find_matching_record_value(
+                ['ip', 'address', 'text', 'destinationHostname'], record)
+            raw_name = self._find_matching_record_value(['host', 'hostname'], record)
+            fqdn_name = self._fqdn_name(raw_name)
+            full_name = self._full_name(raw_name)
+            ptype = rtype if rtype else raw_type
+
+            if name and name not in [raw_name, fqdn_name, full_name]:
+                continue
+
+            if content and content != raw_content:
+                continue
+
+            if rtype and raw_type and rtype != raw_type:
+                continue
+
+            processed_record = {
+                'type': ptype,
+                'name': full_name,
+                'content': raw_content,
+                'ttl': None, # The API does not expose the record's ttl
+                'id': hashlib.md5((ptype+full_name+raw_content).encode('utf-8')).hexdigest()
+            }
+            records.append(processed_record)
+
+        LOGGER.debug('list_records: %s', records)
+        return records
+
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
+        if identifier:
+            records = self._list_records(rtype)
```

Comment:
If this is not too heavy in term of API call, I think you could just get all records in the zone, not filtering on the `rtype`, then find the matching `identifier`. It allows a slightly wider usage of `update_record`.